erpnext/crm/doctype/appointment/appointment.py

```diff
@@ -0,0 +1,215 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2019, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+from __future__ import unicode_literals
+
+import urllib
+from collections import Counter
+from datetime import timedelta
+
+import frappe
+from frappe import _
+from frappe.model.document import Document
+from frappe.utils import get_url
+from frappe.utils.verified_command import verify_request, get_signed_params
+
+
+class Appointment(Document):
+
+    def find_lead_by_email(self):
+        lead_list = frappe.get_list(
+            'Lead', filters={'email_id': self.customer_email}, ignore_permissions=True)
+        if lead_list:
+            return lead_list[0].name
+        return None
+
+    def before_insert(self):
+        number_of_appointments_in_same_slot = frappe.db.count(
+            'Appointment', filters={'scheduled_time': self.scheduled_time})
+        number_of_agents = frappe.db.get_single_value('Appointment Booking Settings', 'number_of_agents')
+        if(number_of_appointments_in_same_slot >= number_of_agents):
+            frappe.throw('Time slot is not available')
+        # Link lead
+        self.lead = self.find_lead_by_email()
+
+    def after_insert(self):
+        if self.lead:
+            # Create Calendar event
+            self.auto_assign()
+            self.create_calendar_event()
+        else:
+            # Set status to unverified
+            self.status = 'Unverified'
+            # Send email to confirm
+            verify_url = self._get_verify_url()
+            template = 'confirm_appointment'
+            args = {
+                "link":verify_url,
+                "site_url":frappe.utils.get_url(),
+                "full_name":self.customer_name,
+            }
+            frappe.sendmail(recipients=[self.customer_email],
+                            template=template,
+                            args=args,
+                            subject=_('Appointment Confirmation'))
+            frappe.msgprint(
+                'Please check your email to confirm the appointment')
+
+
+    def on_change(self):
+        # Sync Calendar
+        if not self.calendar_event:
+            return
+        cal_event = frappe.get_doc('Event', self.calendar_event)
+        cal_event.starts_on = self.scheduled_time
+        cal_event.save(ignore_permissions=True)
+
+
+    def set_verified(self, email):
```

Comment:
When is this function called?

Reply:
The appointment has an "Unverified" state if the lead doesn't already exist in the system. Unverified leads don't have calendar events or leads linked to them. This function creates the necessary leads and events and links them to the appointment. It's called from portal page.